TrackingTools/RecoGeometry/plugins/GlobalDetLayerGeometryESProducer.cc

```diff
@@ -23,11 +23,26 @@ GlobalDetLayerGeometryESProducer::produce(const RecoGeometryRecord & iRecord){
   
   edm::ESHandle<GeometricSearchTracker> tracker;  
   edm::ESHandle<MuonDetLayerGeometry> muon;
+  edm::ESHandle<MTDDetLayerGeometry> mtd;
 
   iRecord.getRecord<TrackerRecoGeometryRecord>().get(tracker);
   iRecord.getRecord<MuonRecoGeometryRecord>().get(muon);
+  
+  // get the MTD if it is available
+  try {
```

Comment:
Is there a way to avoid try/catch? (Less of a problem in ESProducers than EDProducers, but can still make it hard to debug "real" exceptions...)

Reply:
In this case it's catching the specific exception that's thrown when the ES product isn't available because it's not being produced at all(as in the GlobalTrackingGeometryESProducer). Other exceptions are thrown if there's a problem in the creation of the product itself, iirc?

Reply:
Call `iRecord.tryToGetRecord<MTDRecoGeometryRecord>()` instead. That function will return a `nullptr` if the record is missing.

Reply:
Would `tryToGetRecord()` help? I.e. something along

```suggestion
  if(auto mtdRecord = iRecord.tryToGetRecord<MTDRecoGeometryRecord>()) {
     mtdRecord->get(mtd) {
       if(!mtd.isValid()) {
        LogInfo("GlobalDetLayergGeometryBuilder") << "No MTD geometry is available.";
      }
    }
  }
  else {
    LogInfo("GlobalDetLayerGeometryBuilder") << "No MTDDigiGeometryRecord is available.";
  }
```

(note: looking from the framework code I'm not sure if the snippet above really compiles, and the only occurrence of `tryToGetRecord()` that I found is the function template definition)

Reply:
OK I'll give it a try, thanks!

Reply:
In the meanwhile I can confirm that `tryToGetRecord()` doesn't compile at the moment. I'll provide a fix shortly.

Reply:
The fix is in #25094.